### PR TITLE
calicoctl checksystem command

### PR DESCRIPTION
**Pull request by gunjan5**
For issue #85

### CODE REVIEW

File: calicoctl/commands/checksystem.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+			return true
+		}
+	}
+	return false
+}
+
+// checkModule is a utility function used by `checkKernelModules`
+// it opens the file provided and checks if the module passed in
+// as an argument exists for the provided kernelVersion
+func checkModule(filename, module, kernelVersion string) error {
```
Comment (matthewdupre) [inline calicoctl/commands/checksystem.go]: Would it be easier / shorter for this to just be a grep call and check the return code?

File: calicoctl/commands/checksystem.go [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+	modulesLoadablePath := fmt.Sprintf("/lib/modules/%s/modules.dep", kernelVersionStr)
+
+	// File path to Builtin kernel modules
+	modulesBuiltinPath := fmt.Sprintf("/lib/modules/%s/modules.builtin", kernelVersionStr)
+
+	// Keep track of modules that are not found
+	modulesNotFound := []string{}
+
+	// Go through all the required modules and check Loadable and Builtin in order
+	for _, v := range requiredModules {
```
Comment (matthewdupre) [inline calicoctl/commands/checksystem.go]: We could delete the contains() and findAndRemove() functions as well.

File: calicoctl/calicoctl.go [hunk 3]
```diff
@@ -35,6 +35,7 @@ func main() {
     get            Get a resource identified by file, stdin or resource type and name.
     version        Display the version of calicoctl.
     node           Node related commands.
+    checksystem    Check for incompatibilities on the host system.
```
Comment (gunjan5) [inline calicoctl/calicoctl.go]: yup! I had it worded like that before, but then I realized python version has it that way so I changed it. I'll change it back to more positive

File: calicoctl/commands/checksystem.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+)
+
+// The minimum allowed linux kernel version is 2.6.24, which introduced network
+// namespaces and veth pairs.
+const minKernelVersion = "2.6.24"
+
+// Required kernel modules to run Calico
+var requiredModules = []string{"xt_set", "ip6_tables"}
+
+// Checksystem checks host system for compatible versions
```
Comment (gunjan5) [inline calicoctl/commands/checksystem.go]: Comment for any exported function in Go has to begin with it's name for godocs to work, that's why it's like `// Checksystem ...`.

Versions being validated are listed in comments (line 31,32)

File: calicoctl/commands/checksystem.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+func Checksystem(args []string) error {
+	doc := `Usage: 
+    calicoctl checksystem
+
+Description:
+    Check for incompatibilities between Calico and the host system`
+
+	parsedArgs, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
```
Comment (gunjan5) [inline calicoctl/commands/checksystem.go]: - `calicoctl` in version is to be consistent with all the other commands in `libcalico-go` for now. For now all the `calicoctl` commands print `calicoctl` when you pass `--version` 

- `docopts` port in Go is very hacky, but it is hittable. When `-h` is passed it treats it as no args passed O.o (as explained in the comment below)

File: calicoctl/commands/checksystem.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+		fmt.Printf("%-20s\t\t\t%s", "System doesn't meet one or more minimum systems requirements to run Calico", "\u2717\n")
+		os.Exit(1)
+	}
+
+	fmt.Printf("%-20s\t\t\t%s", "System meets minimum system requirements to run Calico!", "\u2713\n")
+
+	return nil
+}
+
+// checkKernelVersion checks for minimum required kernel version
```
Comment (gunjan5) [inline calicoctl/commands/checksystem.go]: same as before, godocs require the comment to begin with func/field name, and version being verified is listed at the top in comment, writing it here would be redundant (also someone might forget to update it in future)

File: calicoctl/commands/checksystem.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+Description:
+    Check for incompatibilities between Calico and the host system`
+
+	parsedArgs, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
+	if err != nil {
+		return err
+	}
+
+	// Return if no args are passed or `-h` `--help` is passed
```
Comment (gunjan5) [inline calicoctl/commands/checksystem.go]: as I'm reading it now, it does sound strange :) I'll try to reword it

File: calicoctl/commands/checksystem.go [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+	// Print the version for the command output
+	fmt.Printf("\t\t%-20s", kernelVersionStr)
+
+	// Goversion normalizes the versions and compares them returns `true` if
+	// running version is >= minimum version
+	if !goversion.CompareNormalized(kernelVersionStr, minKernelVersion, ">=") {
+
+		// Prints "✗" if current version is not >= minimum required version
+		fmt.Printf("\t\t\t\t\t\u2717\n")
```
Comment (gunjan5) [inline calicoctl/commands/checksystem.go]: you're right, didn't think about telnet. I'll replace ✓ with `OK.` and ✗ with `FAIL`

File: calicoctl/commands/checksystem.go (hunk 2 above)
Comment (gunjan5) [inline calicoctl/commands/checksystem.go]: ok restructured it into a single loop but in a way `lsmod` is called only once

File: calicoctl/commands/checksystem.go [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+	docopt "github.com/docopt/docopt-go"
+	goversion "github.com/mcuadros/go-version"
+)
+
+// The minimum allowed linux kernel version is 2.6.24, which introduced network
+// namespaces and veth pairs.
+const minKernelVersion = "2.6.24"
+
+// Required kernel modules to run Calico
+var requiredModules = []string{"xt_set", "ip6_tables"}
```
Comment (gunjan5) [inline calicoctl/commands/checksystem.go]: fixed!

File: calicoctl/commands/checksystem.go (hunk 4 above)
Comment (matthewdupre) [inline calicoctl/commands/checksystem.go]: Silly godocs, but fair enough.

File: calicoctl/commands/checksystem.go (hunk 5 above)
Comment (matthewdupre) [inline calicoctl/commands/checksystem.go]: "calicoctl" is not valid for --version; we need to fix this (everywhere) before it goes 1.0.  You can leave it and raise an issue if you like, or fix them (all ideally) here.

Comment (matthewdupre) [inline calicoctl/commands/checksystem.go]: It only squelches the -h because we tell it to in the Parse() call.

Comment (matthewdupre) [inline calicoctl/commands/checksystem.go]: (I'll reread this method once I've been through all the discussion, I really need to just look at it.)

File: calicoctl/commands/node/checksystem.go [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+	fmt.Print("Checking kernel modules...\n")
+	err = checkKernelModules()
+	if err != nil {
+		systemOk = false
+	}
+
+	// If any of the checks fail, print a message and exit
+	if !systemOk {
+		fmt.Printf("%-20s\t\t\t%s", "System doesn't meet one or more minimum systems requirements to run Calico", "FAIL\n")
```
Comment (matthewdupre) [inline calicoctl/commands/node/checksystem.go]: The `-20` is superfluous here?

File: calicoctl/commands/node/checksystem.go [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+		} else {
+			fmt.Printf("\t\t%-20s%s", v, "\t\t\t\t\tOK\n")
+		}
+	}
+
+	// If there are still any modules not found then return an error
+	if len(modulesNotFound) > 0 {
+
+		// ip6_tables is not a required module for ipv4 setups, so just print
+		// a warning instead of failing the systemc heck
```
Comment (matthewdupre) [inline calicoctl/commands/node/checksystem.go]: Space still in wrong place.

File: calicoctl/commands/node/checksystem.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+		return errors.New("One of more kernel modules missing")
+	}
+
+	return nil
+}
+
+// checkModule is a utility function used by `checkKernelModules`
+// it opens the file provided and checks if the module passed in
+// as an argument exists for the provided kernelVersion
+func checkModule(filename, module, kernelVersion string) error {
```
Comment (matthewdupre) [inline calicoctl/commands/node/checksystem.go]: I still don't see the point of having a page of code to replace a single OS command, but if you really want it (and the unit testing etc. responsibilities that will go with this), then fine.

File: calicoctl/commands/node/checksystem.go (hunk 10 above)
Comment (gunjan5) [inline calicoctl/commands/node/checksystem.go]: it is so the `OK` and `FAIL` line up with other `OK` and `FAIL`s

File: calicoctl/commands/checksystem.go [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+		}
+	}
+
+	// If there are still any modules not found then return an error
+	if len(modulesNotFound) > 0 {
+
+		// ip6_tables is not a required module for ipv4 setups, so just print
+		// a warning instead of failing the systemc heck
+		if len(modulesNotFound) == 1 && modulesNotFound[0] == "ip6_tables" {
+			fmt.Printf("WARNING: Unable to detect ip6_tables kernel module")
```
Comment (gunjan5) [inline calicoctl/commands/checksystem.go]: fixed

**Pull request merged**